pkg/operator/controller/certificate/publish_certs.go

```diff
@@ -0,0 +1,160 @@
+		}
+	case desired != nil && current != nil:
+		if updated, err := r.updateRouterCertsGlobalSecret(current, desired); err != nil {
+			return fmt.Errorf("failed to update published router certificates secret: %v", err)
+		} else if updated {
+			r.recorder.Eventf(desired, "Normal", "UpdatedPublishedRouterCertificates", "Updated the published router certificates")
+		}
+	}
+	return nil
+}
+
+// desiredRouterCertsGlobalSecret returns the desired router-certs global
+// secret.
+func desiredRouterCertsGlobalSecret(secrets []corev1.Secret, ingresses []ingressv1alpha1.ClusterIngress) (*corev1.Secret, error) {
+	if len(ingresses) == 0 || len(secrets) == 0 {
+		return nil, nil
+	}
+
+	nameToSecret := map[string]*corev1.Secret{}
+	for i, certSecret := range secrets {
+		nameToSecret[certSecret.Name] = &secrets[i]
+	}
+
+	ingressToSecret := map[*ingressv1alpha1.ClusterIngress]*corev1.Secret{}
+	for i, ingress := range ingresses {
+		secretName := fmt.Sprintf("router-certs-%s", ingress.Name)
+		if ingress.Spec.DefaultCertificateSecret != nil && len(*ingress.Spec.DefaultCertificateSecret) > 0 {
+			secretName = *ingress.Spec.DefaultCertificateSecret
+		}
+		if secret, ok := nameToSecret[secretName]; ok {
+			ingressToSecret[&ingresses[i]] = secret
+		}
+	}
+
+	name := controller.RouterCertsGlobalSecretName()
+	globalSecret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name.Name,
+			Namespace: name.Namespace,
+		},
+		Data: map[string][]byte{},
+	}
+	for ingress, certSecret := range ingressToSecret {
+		if len(ingress.Status.IngressDomain) == 0 {
+			continue
+		}
+		pem := bytes.Join([][]byte{
+			certSecret.Data["tls.crt"],
+			certSecret.Data["tls.key"],
+		}, nil)
+		globalSecret.Data[ingress.Status.IngressDomain] = pem
+	}
+	return globalSecret, nil
+}
+
+// currentRouterCertsGlobalSecret returns the current router-certs global
+// secret.
+func (r *reconciler) currentRouterCertsGlobalSecret() (*corev1.Secret, error) {
+	name := controller.RouterCertsGlobalSecretName()
+	secret := &corev1.Secret{}
+	if err := r.client.Get(context.TODO(), name, secret); err != nil {
+		if errors.IsNotFound(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+	return secret, nil
+}
+
+// createRouterCertsGlobalSecret creates a router-certs global secret.  Returns
+// true if the secret was created, false otherwise.
+func (r *reconciler) createRouterCertsGlobalSecret(secret *corev1.Secret) (bool, error) {
+	if err := r.client.Create(context.TODO(), secret); err != nil {
+		if errors.IsAlreadyExists(err) {
+			return false, nil
+		}
+		return false, err
+	}
+	return true, nil
+}
+
+// updateRouterCertsGlobalSecret updates the router-certs global secret.
+// Returns true if the secret was updated, false otherwise.
+func (r *reconciler) updateRouterCertsGlobalSecret(current, desired *corev1.Secret) (bool, error) {
+	if routerCertsSecretsEqual(current, desired) {
+		return false, nil
+	}
+	updated := current.DeepCopy()
```

Comment:
If the update is successful, should we update into `current` instead of a copy? 😁

Reply:
Actually, if we _don't_ update into `current`, haven't we invalidated future updates of `current` (because `current` becomes a stale resource version)?

Reply:
> Actually, if we _don't_ update into `current`, haven't we invalidated future updates of `current` (because `current` becomes a stale resource version)?

In this case, it shouldn't matter because the extent terminates when `ensureRouterCertsGlobalSecret` returns.  However, I do want to set a good example.  I wonder whether we want to establish a pattern of updating into the current value, or whether we want to establish a pattern of functions' returning a Boolean value indicating whether they changed the object, leaving it up to the caller to get the new version if needed.

Reply:
Maybe just doc it for now, or mutate and doc?

Reply:
Can address in a followup

Reply:
Current godoc says, "Returns true if the secret was updated, false otherwise." We could add "Caller is responsible for getting the updated secret if needed."